File: class/class5.1/src/MegaDataPool.cpp
#include "MegaDataPool.hpp"

MegaDataPool::MegaDataPool(size_t poolSize)
    : poolSize(poolSize), available(poolSize), used(0) {
    for (size_t i = 0; i < poolSize; ++i) {
        pool.emplace_back();
        availableIndices.push(i);
    }    
}

MegaData& MegaDataPool::acquire() {
    if (available == 0){
        throw std::runtime_error("No available MegaData objects in the pool");
    }

    size_t index = availableIndices.front();
    availableIndices.pop();
    --available;
    ++used;

    return pool[index];
}

void MegaDataPool::release(MegaData& data) {
    auto it = std::ranges::find_if(pool, [&data](const MegaData& d) { return &d == &data; });
    if (it != pool.end()) {
        size_t index = std::distance(pool.begin(), it);
        availableIndices.push(index);
        --used;
        ++available;
    }
}

size_t MegaDataPool::size() const {
    return this->poolSize;
}
size_t MegaDataPool::usedSize() const {
    return this->used;
}
size_t MegaDataPool::availableSize() const {
    return this->available;
}

File: home/home5.1/object-pool/src/mega_data_pool.rs
use crate::mega_data::MegaData;

pub struct MegaDataPool {
    pub pool: Vec<MegaData>,
    pub available_indices: Vec<usize>,
}

impl MegaDataPool {
    pub fn new(pool_size: usize) -> Self {
        let mut pool = Vec::with_capacity(pool_size);
        let mut available_indices = Vec::with_capacity(pool_size);

        for i in 0..pool_size {
            pool.push(MegaData::new());
            available_indices.push(i);
        }

        Self {
            pool,
            available_indices
        }
    }

    pub fn acquire(&mut self) -> Option<&mut MegaData> {
        if let Some(index) = self.available_indices.pop() {
            Some(&mut self.pool[index])
        }
        else {
            None
        }
    }

    pub fn release(&mut self, data: &mut MegaData) {
        let index = self.pool.iter_mut().position(|d| std::ptr::eq(d, data));
        if let Some(index) = index {
            self.available_indices.push(index);
        }
    }

    pub fn size(&self) -> usize {
        self.pool.len()
    }

    pub fn used_size(&self) -> usize {
        self.pool.len() - self.available_indices.len()
    }
}
